### PR TITLE
SortingHat server

### PR DESCRIPTION
**Pull request by sduenas**
This PR adds a first version of the SortingHat server script. By default, the server runs a WSGI app because it is supposed to be deployed with a reverse HTTP server in production. However, it can run a HTTP server when the `--dev` flag is activated. When this flag is active, the server will also serve the UI interface. To test it, you should follow the next steps.

1. Install the new dependencies with poetry install
```
$ poetry install
```

2. Build the UI interface artifact. The static files will be copied to `sortinghat/core/static` and `sortinghat/core/templates`.
```
$ cd ui
$ yarn build
```

3. Copy the static files to its destination
```
$ ./manage.py collectstatic --settings=config.settings.devel
```

4. Run the server
```
$ sortinghatd --config config.settings.devel --dev
```

The url http://localhost:8080 should show the login page.

To test it with a reverse proxy like NGINX using docker-compose the next files are necesary:

- uwsig_params
```
uwsgi_param QUERY_STRING $query_string;
uwsgi_param REQUEST_METHOD $request_method;
uwsgi_param CONTENT_TYPE $content_type;
uwsgi_param CONTENT_LENGTH $content_length;
uwsgi_param REQUEST_URI $request_uri;
uwsgi_param PATH_INFO $document_uri;
uwsgi_param DOCUMENT_ROOT $document_root;
uwsgi_param SERVER_PROTOCOL $server_protocol;
uwsgi_param HTTPS $https if_not_empty;
uwsgi_param REMOTE_ADDR $remote_addr;
uwsgi_param REMOTE_PORT $remote_port;
uwsgi_param SERVER_PORT $server_port;
uwsgi_param SERVER_NAME $server_name;
```

- nginx.cfg
```
events { }
http {
  upstream sortinghat {
      server 172.17.0.1:6314;
  }
  server {
      include    mime.types;
      sendfile on;
      listen 8080;
      server_name localhost;
  
      location / {
          include /etc/nginx/uwsgi_params;
          uwsgi_pass sortinghat;
          uwsgi_param Host $host;
          uwsgi_param X-Real-IP $remote_addr;
          uwsgi_param X-Forwarded-For $proxy_add_x_forwarded_for;
          uwsgi_param X-Forwarded-Proto $http_x_forwarded_proto;
      }

      location /static {
        root /tmp;
      }
      location /css {
        root /tmp/static;
      }
      location /js {
        root /tmp/static;
      }

      location /fonts {
        root /tmp/static;
      }
  }
}
```
- nginx.yml
```
version: '3'

services:
  nginx:
    image: nginx:latest
    container_name: webserver
    restart: unless-stopped
    ports:
      - 8080:8080
    volumes:
      - /tmp/nginx.conf:/etc/nginx/nginx.conf
      - /tmp/uwsgi_params:/etc/nginx/uwsgi_params
      - /tmp/static:/tmp/stati
```

The run the server with the next command:
```
$ sortinghatd --config config.settings.devel
```

And run docker compose with:
```
docker-compose -f nginx.yml up
```

The same url should show the login page.

### CODE REVIEW

**Comment by ajaragz** [comment]
@sduenas it works fine. Maybe we need to update the settings file with the changes below. There's other changes that could be updated in the PR description. Other than that it LGTM

-------

Changes I had to make to the settings
```diff
diff --git a/config/settings/devel.py b/config/settings/devel.py
index 74c85ed..85d1702 100644
--- a/config/settings/devel.py
+++ b/config/settings/devel.py
@@ -85,13 +85,13 @@ SECRET_KEY = 'fake-key'
 # https://docs.djangoproject.com/en/3.2/howto/static-files/
 
 STATICFILES_DIRS = [
-    "../sortinghat/sortinghat/core/static",
+    "./sortinghat/core/static",
 ]
 STATIC_ROOT = "/tmp/static/"
 
 STATIC_URL = '/'
 
-MEDIA_URL = 'media.example.com/'
+MEDIA_URL = 'http://media.example.com/'
 
 INSTALLED_APPS = [
     'django_rq',
@@ -109,6 +109,7 @@ INSTALLED_APPS = [
 DATABASES = {
     'default': {
         'ENGINE': 'django.db.backends.mysql',
+        'HOST': '127.0.0.1',
         'USER': 'root',
         'PASSWORD': '',
         'NAME': 'sortinghat_db',

```
After a MariaDB container is up, commands to have the DB backend ready
```shell
sudo apt install libmariadbclient-dev
poetry install
poetry shell
./manage.py makemigrations --settings=config.settings.devel
./manage.py migrate --settings=config.settings.devel
./manage.py loaddata sortinghat/core/fixtures/countries.json --settings=config.settings.devel
./manage.py createsuperuser --settings=config.settings.devel
```

Generate UI and copy as static files
```shell
cd ui
yarn add @vue/cli
yarn build
cd ..
./manage.py collectstatic --settings=config.settings.devel
```
> NOTE: I had to install the Vue cli to be able to build the UI


Server run in dev mode
----------------------

Started with
```shell
sortinghatd --config config.settings.devel --dev
```
I can use UI in http://localhost:8000 (**not 8080**). The GraphQL API also works in http://localhost:8000/api

Server run in production mode
-----------------------------

Compose for NGINX
```dockerfile
version: '3'

services:
  nginx:
    image: nginx:latest
    container_name: webserver
    restart: unless-stopped
    ports:
      - 8080:8080
    volumes:
      - ./nginx.conf:/etc/nginx/nginx.conf
      - ./uwsgi_params:/etc/nginx/uwsgi_params
      - /tmp/static:/tmp/static

```
> NOTE: I had to use relative paths for the NGINX config files and fix typos in their names from the PR description: `nginx.cfg` to `nginx.conf` and `uwsig_params` to `uwsgi_params`. Also a typo in `/tmp/stati`

After NGINX container is running, server is launched with
```shell
sortinghatd --config config.settings.devel
```

I can use the UI in http://localhost:8080

> NOTE: no API in http://localhost:8080/api however

File: README.md [hunk 1]
```diff
@@ -129,6 +130,45 @@ Run SortingHat frontend Vue app:
 $ yarn serve
 ```
 
+
+## SortingHat service
+
+Starting on version 0.8, SortingHat is released with a server app. The server has two
```
Comment (ajaragz) [inline README.md:136]: This should  be _starting **at** version 0.8_

File: README.md [hunk 2]
```diff
@@ -129,6 +130,45 @@ Run SortingHat frontend Vue app:
...
+modes, `production` and `development`.
+
+When `production` mode is active, a WSGI app is served. The idea is to use a reverse
+proxy like NGINX or similar, that will be connected with the WSGI app to provide
+an interface HTTP.
+
+When `development` mode is active, a HTTP server is launched, so you can interact
```
Comment (ajaragz) [inline README.md:143]: _a**n**_ HTTP server

Comment (sduenas) [reply README.md]: Fixed.

File: README.md (hunk 1 above)
Comment (sduenas) [reply README.md]: Fixed.

File: README.md [hunk 3]
```diff
@@ -129,6 +130,45 @@ Run SortingHat frontend Vue app:
...
+
+You will need a django configuration file to run the service. You can use and modify
+one of the ones stored in `config/settings` folder. The file must be accessible
+via `PYTHONPATH` env variable. For the next examples, we will use `devel.py` file.
+
+In order to run the service for the first time, you need to execute the next commands:
+
+Build the UI interface:
+```
+$ yarn build
```
Comment (ajaragz) [inline README.md:155]: this should include `cd ui` to know from where to run the yarn command

Comment (sduenas) [reply README.md:156]: Fixed.

**Comment by ajaragz** [comment]
LGTM

**Review by ajaragz — APPROVED** [review]

**Pull request merged**